File: src/main/java/xyz/oribuin/lilori/LilOri.java
package xyz.oribuin.lilori;

import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import xyz.oribuin.lilori.commands.CmdCommand;
import xyz.oribuin.lilori.commands.CmdHelp;
import xyz.oribuin.lilori.commands.CmdPing;
import xyz.oribuin.lilori.commands.CmdPrefix;
import xyz.oribuin.lilori.commands.administrative.CmdPerms;
import xyz.oribuin.lilori.commands.author.CmdEval;
import xyz.oribuin.lilori.commands.author.CmdQuery;
import xyz.oribuin.lilori.commands.author.CmdShutdown;
import xyz.oribuin.lilori.commands.author.CmdTest;
import xyz.oribuin.lilori.commands.games.*;
import xyz.oribuin.lilori.commands.moderation.CmdBan;
import xyz.oribuin.lilori.commands.moderation.CmdKick;
import xyz.oribuin.lilori.commands.moderation.CmdMute;
import xyz.oribuin.lilori.commands.moderation.CmdPurge;
import xyz.oribuin.lilori.commands.music.CmdPlay;
import xyz.oribuin.lilori.commands.music.CmdStop;
import xyz.oribuin.lilori.database.DatabaseConnector;
import xyz.oribuin.lilori.database.SQLiteConnector;
import xyz.oribuin.lilori.listeners.EventMentionOri;
import xyz.oribuin.lilori.listeners.Presence;
import xyz.oribuin.lilori.managers.GuildSettingsManager;
import xyz.oribuin.lilori.managers.command.Command;
import xyz.oribuin.lilori.managers.command.CommandExecutor;
import xyz.oribuin.lilori.managers.command.CommandHandler;
import xyz.oribuin.lilori.utils.EventWaiter;

import javax.security.auth.login.LoginException;
import java.io.File;
import java.io.IOException;

public class LilOri extends ListenerAdapter {

    private EventWaiter waiter = new EventWaiter();
    private static LilOri instance;
    private DatabaseConnector connector;
    private GuildSettingsManager guildSettingsManager;
    private CommandHandler commandHandler;

    private LilOri() throws LoginException {
        instance = this;

        // Setup the SQLite Database
        File file = new File("data", "lilori.db");
        try {
            if (!file.exists()) {
                file.createNewFile();

                System.out.println("Created SQLite Database File: lilori.db");
            }

            // Register SQLite Connector
            this.connector = new SQLiteConnector(file);
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Setup Managers
        this.guildSettingsManager = new GuildSettingsManager(this);
        this.commandHandler = new CommandHandler();

        this.registerCommands();
        enable();

        // Login Bot
        JDA jda = JDABuilder.createDefault(Settings.TOKEN)
                .addEventListeners(waiter, new CommandExecutor(this, commandHandler),
                        new Presence(),
                        new EventMentionOri()
                ).build();


        System.out.println("*=* Loading Lil' Ori Commands *=*");
        // Startup Message
        int i = 0;
        for (Command command : this.getCommandHandler().getCommands()) {

            System.out.println("Loaded Command: " + command.getName() + " | (" + ++i + "/" + this.getCommandHandler().getCommands().size() + ")");
        }

        System.out.println("*=* Loaded Up " + jda.getSelfUser().getName() + " with " + this.getCommandHandler().getCommands().size() + " Command(s) *=*");
    }

    public static void main(String... args) {
        try {
            new LilOri();
        } catch (LoginException e) {
            e.printStackTrace();
        }
    }

    public static LilOri getInstance() {
        return instance;
    }

    private void registerCommands() {
        this.getCommandHandler().registerCommands(
                // General Commands
                new CmdHelp(),
                new CmdPing(),
                new CmdPrefix(),
                new CmdCommand(),

                // Music Commands
                new CmdPlay(),
                new CmdStop(),

                // Game Commands
                new CmdCoinflip(),
                new CmdColor(),
                new CmdEightball(),
                new CmdFeed(),
                new CmdGay(),
                new CmdQuote(),
                new CmdSlap(),

                // Moderation
                new CmdBan(),
                new CmdKick(),
                new CmdMute(),
                new CmdPurge(this.waiter),

                // Author
                new CmdEval(),
                new CmdQuery(),
                new CmdShutdown(),
                new CmdTest(),

                // Admin
                new CmdPerms()
        );
    }

    private void enable() {
        this.guildSettingsManager.enable();
    }

    public DatabaseConnector getConnector() {
        return connector;
    }

    public GuildSettingsManager getGuildSettingsManager() {
        return guildSettingsManager;
    }

    public CommandHandler getCommandHandler() {
        return commandHandler;
    }
}

File: src/main/kotlin/xyz/oribuin/lilori/LilOri.kt
package xyz.oribuin.lilori

import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.requests.GatewayIntent
import xyz.oribuin.lilori.database.DatabaseConnector
import xyz.oribuin.lilori.database.SQLiteConnector
import xyz.oribuin.lilori.handler.CommandExecutor
import xyz.oribuin.lilori.handler.CommandHandler
import xyz.oribuin.lilori.listener.GeneralEvents
import xyz.oribuin.lilori.listener.support.SupportListeners
import xyz.oribuin.lilori.manager.*
import xyz.oribuin.lilori.util.ConsoleColors
import xyz.oribuin.lilori.util.EventWaiter
import xyz.oribuin.lilori.util.FileUtils
import javax.security.auth.login.LoginException
import kotlin.reflect.KClass


class LilOri : ListenerAdapter() {
    private val managers = mutableMapOf<KClass<out Manager>, Manager>()
    var jdabot: JDA? = null

    // Define handlers
    var connector: DatabaseConnector

    // Define others
    val eventWaiter = EventWaiter()

    init {
        instance = this
        // Setup the SQLite Database
        connector = SQLiteConnector(FileUtils.createFile("data", "lilori.db"))


        // Setup Managers
        getManager(DataManager::class)
        getManager(BotManager::class)
        getManager(CommandHandler::class)
        getManager(EconomyManager::class)
        getManager(GuildSettingsManager::class)
        getManager(QuoteManager::class)
        getManager(TicketManager::class)

        this.managers.values.forEach { manager -> manager.enable() }

        // Register plugin commands
        this.getManager(CommandHandler::class).registerCommands()

        // Login Bot
        val jda = JDABuilder.create(Settings.TOKEN, GatewayIntent.values().toList())
                .addEventListeners(CommandExecutor(this, getManager(CommandHandler::class)), GeneralEvents(this), SupportListeners(), eventWaiter, this)
                .enableIntents(GatewayIntent.values().toList())

        this.jdabot = jda.build()

        // Get the BotManager
        val botManager = getManager(BotManager::class)

        // Register values specifically
        botManager.registerGuilds(jdabot!!)
        botManager.registerStatus(jdabot!!)

        // Startup Log
        this.logStartup(jdabot!!)
    }

    companion object {
        @JvmStatic
        lateinit var instance: LilOri

        @JvmStatic
        fun main(args: Array<String>) {
            try {
                LilOri()
            } catch (e: LoginException) {
                e.printStackTrace()
            }
        }

    }

    fun <M : Manager> getManager(managerClass: KClass<M>): M {
        synchronized(this.managers) {
            @Suppress("UNCHECKED_CAST")
            if (this.managers.containsKey(managerClass))
                return this.managers[managerClass] as M

            return try {
                val manager = managerClass.constructors.first().call(this)
                manager.enable()
                this.managers[managerClass] = manager
                manager
            } catch (ex: ReflectiveOperationException) {
                error("Failed to load manager for ${managerClass.simpleName}")
            }
        }
    }

    private fun logStartup(jdaBot: JDA) {

        // Startup Command Log
        println(ConsoleColors.BLUE_BOLD_BRIGHT + "*=* Loading Lil' Ori Commands *=*" + ConsoleColors.RESET)
        var i = 0

        // Add every command into the console with a number
        for (command in getManager(CommandHandler::class).commands)
            println(ConsoleColors.BLUE_BRIGHT + "Loaded Command: (${command.getAnnotation(command.javaClass).category.categoryName}) ${command.getAnnotation(command.javaClass).name} | (${++i}/${getManager(CommandHandler::class).commands.size})" + ConsoleColors.RESET)

        println(ConsoleColors.GREEN_UNDERLINED + "*=* Loaded Up ${jdaBot.selfUser.name} with ${getManager(CommandHandler::class).commands.size}  Command(s) *=*" + ConsoleColors.RESET)
    }
}